File: daggercore/src/main/java/com/worldsnas/daggercore/navigator/changehandler/FlipChangeHandler.java
package com.worldsnas.daggercore.navigator.changehandler;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.util.Property;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import com.bluelinelabs.conductor.changehandler.AnimatorChangeHandler;

public class FlipChangeHandler extends AnimatorChangeHandler {

    private static final long DEFAULT_ANIMATION_DURATION = 300;

    public enum FlipDirection {
        LEFT(-180, 180, View.ROTATION_Y),
        RIGHT(180, -180, View.ROTATION_Y),
        UP(-180, 180, View.ROTATION_X),
        DOWN(180, -180, View.ROTATION_X);

        final int inStartRotation;
        final int outEndRotation;
        final Property<View, Float> property;

        FlipDirection(int inStartRotation, int outEndRotation, Property<View, Float> property) {
            this.inStartRotation = inStartRotation;
            this.outEndRotation = outEndRotation;
            this.property = property;
        }
    }

    private final long animationDuration;
    private final FlipDirection flipDirection;

    @Keep
    public FlipChangeHandler() {
        this(FlipDirection.RIGHT);
    }

    public FlipChangeHandler(FlipDirection flipDirection) {
        this(flipDirection, DEFAULT_ANIMATION_DURATION);
    }

    public FlipChangeHandler(long animationDuration) {
        this(FlipDirection.RIGHT, animationDuration);
    }

    public FlipChangeHandler(FlipDirection flipDirection, long animationDuration) {
        this.flipDirection = flipDirection;
        this.animationDuration = animationDuration;
    }

    @Override @NonNull
    protected Animator getAnimator(@NonNull ViewGroup container, View from, View to, boolean isPush, boolean toAddedToContainer) {
        AnimatorSet animatorSet = new AnimatorSet();

        if (to != null) {
            to.setAlpha(0);

            ObjectAnimator rotation = ObjectAnimator.ofFloat(to, flipDirection.property, flipDirection.inStartRotation, 0).setDuration(animationDuration);
            rotation.setInterpolator(new AccelerateDecelerateInterpolator());
            animatorSet.play(rotation);

            Animator alpha = ObjectAnimator.ofFloat(to, View.ALPHA, 1).setDuration(animationDuration / 2);
            alpha.setStartDelay(animationDuration / 3);
            animatorSet.play(alpha);
        }

        if (from != null) {
            ObjectAnimator rotation = ObjectAnimator.ofFloat(from, flipDirection.property, 0, flipDirection.outEndRotation).setDuration(animationDuration);
            rotation.setInterpolator(new AccelerateDecelerateInterpolator());
            animatorSet.play(rotation);

            Animator alpha = ObjectAnimator.ofFloat(from, View.ALPHA, 0).setDuration(animationDuration / 2);
            alpha.setStartDelay(animationDuration / 3);
            animatorSet.play(alpha);
        }

        return animatorSet;
    }

    @Override
    protected void resetFromView(@NonNull View from) {
        from.setAlpha(1);

        switch (flipDirection) {
            case LEFT:
            case RIGHT:
                from.setRotationY(0);
                break;
            case UP:
            case DOWN:
                from.setRotationX(0);
                break;
        }
    }
}

File: navigation/src/androidMain/kotlin/com/worldsnas/navigation/changehandler/FlipChangeHandler.kt
package com.worldsnas.navigation.changehandler

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.util.Property
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import androidx.annotation.Keep
import com.bluelinelabs.conductor.changehandler.AnimatorChangeHandler

class FlipChangeHandler @JvmOverloads constructor(
    private val flipDirection: FlipDirection,
    private val animationDurationMillis: Long = DEFAULT_ANIMATION_DURATION
) : AnimatorChangeHandler() {
    enum class FlipDirection(
        val inStartRotation: Int,
        val outEndRotation: Int,
        val property: Property<View, Float>
    ) {
        LEFT(-180, 180, View.ROTATION_Y), RIGHT(
            180,
            -180,
            View.ROTATION_Y
        ),
        UP(-180, 180, View.ROTATION_X), DOWN(180, -180, View.ROTATION_X);

    }

    @Keep
    constructor() : this(FlipDirection.RIGHT) {
    }

    constructor(animationDuration: Long) : this(FlipDirection.RIGHT, animationDuration) {}

    override fun getAnimator(
        container: ViewGroup,
        from: View?,
        to: View?,
        isPush: Boolean,
        toAddedToContainer: Boolean
    ): Animator {
        val animatorSet = AnimatorSet()
        if (to != null) {
            to.alpha = 0f
            val rotation =
                ObjectAnimator.ofFloat(
                    to,
                    flipDirection.property,
                    flipDirection.inStartRotation.toFloat(),
                    0f
                ).setDuration(animationDurationMillis)
            rotation.interpolator = AccelerateDecelerateInterpolator()
            animatorSet.play(rotation)
            val alpha: Animator =
                ObjectAnimator.ofFloat(
                    to,
                    View.ALPHA,
                    1f
                ).setDuration(animationDurationMillis / 2)
            alpha.startDelay = animationDurationMillis / 3
            animatorSet.play(alpha)
        }
        if (from != null) {
            val rotation =
                ObjectAnimator.ofFloat(
                    from,
                    flipDirection.property,
                    0f,
                    flipDirection.outEndRotation.toFloat()
                ).setDuration(animationDurationMillis)
            rotation.interpolator = AccelerateDecelerateInterpolator()
            animatorSet.play(rotation)
            val alpha: Animator =
                ObjectAnimator.ofFloat(
                    from,
                    View.ALPHA,
                    0f
                ).setDuration(animationDurationMillis / 2)
            alpha.startDelay = animationDurationMillis / 3
            animatorSet.play(alpha)
        }
        return animatorSet
    }

    override fun resetFromView(from: View) {
        from.alpha = 1f
        when (flipDirection) {
            FlipDirection.LEFT, FlipDirection.RIGHT -> from.rotationY = 0f
            FlipDirection.UP, FlipDirection.DOWN -> from.rotationX = 0f
        }
    }

    companion object {
        private const val DEFAULT_ANIMATION_DURATION: Long = 300
    }

}